In [10]:
import sympy as sp
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression

# Домашнее задание

Математический анализ
69 баллов

**Задача 1** (14 баллов)

Найдите экстремумы функции:

$$
f(x, y, z)=2 x^3+2 x y+2 x z+y^2+z^2+2 y-8 .
$$

Распишите подробное решение.

In [11]:
# переменные
x, y, z = sp.symbols('x, y, z')

# функция
f = 2 * x**3 + 2 * x * y + 2 * x * z + y**2 + z**2 + 2 * y - 8

# частные производные
df_dx = f.diff(x)
df_dy = f.diff(y)
df_dz = f.diff(z)

print("Частные производные:")
print(f"df/dx = {df_dx}")
print(f"df/dy = {df_dy}")
print(f"df/dz = {df_dz}")

#система уравнений (производные = 0)
eq = [sp.Eq(df_dx, 0), sp.Eq(df_dy, 0), sp.Eq(df_dz, 0)]
print('\nСистема уравнений:', *eq, sep='\n')

# приравнивание производных к 0 и решение системы
solutions = sp.solve(eq, [x, y, z])
print('\nРешения (x, y, z):', *solutions, sep='\n')

# анализ второго порядка
df_dxdx = df_dx.diff(x)
df_dydy = df_dy.diff(y)
df_dzdz = df_dz.diff(z)
df_dxdy = df_dx.diff(y)
df_dxdz = df_dx.diff(z)
df_dydz = df_dy.diff(z)

print("Производные второго порядка:")
print(f"df/dxdx = {df_dxdx}")
print(f"df/dydy = {df_dydy}")
print(f"df/dzdz = {df_dzdz}")
print(f"df/dxdy = {df_dxdy}")
print(f"df/dxdz = {df_dxdz}")
print(f"df/dydz = {df_dydz}")

Частные производные:
df/dx = 6*x**2 + 2*y + 2*z
df/dy = 2*x + 2*y + 2
df/dz = 2*x + 2*z

Система уравнений:
Eq(6*x**2 + 2*y + 2*z, 0)
Eq(2*x + 2*y + 2, 0)
Eq(2*x + 2*z, 0)

Решения (x, y, z):
(-1/3, -2/3, 1/3)
(1, -2, -1)
Производные второго порядка:
df/dxdx = 12*x
df/dydy = 2
df/dzdz = 2
df/dxdy = 2
df/dxdz = 2
df/dydz = 0


In [12]:
# функция для вычисления матрицы Гессе (только x т.к. только он остался в прозводных второго порядка)
def find_hesse(x):
    return np.array([
        [12 * x, 2, 2],
        [2, 2, 0],
        [2, 0, 2]
    ], dtype=float)

# проверка экстремумов
for point in solutions:
    x = point[0]
    # счет матрицы Гессе для x
    hesse = find_hesse(x)
    # определитель
    hesse_det = np.linalg.det(hesse)
    # собственные числа
    eig = np.linalg.eigvals(hesse)

    # определение типа экстремума
    if hesse_det > 0 and eig.all() > 0:
        print(f'Минимум в точке {point}')
    elif hesse_det > 0 and eig.all() < 0:
        print(f'Максимум в точке {point}')
    elif hesse_det < 0:
        print(f'Седловая точка в точке {point}')
    else:
        print(f'{point} не является точкой экстремума')

Седловая точка в точке (-1/3, -2/3, 1/3)
Минимум в точке (1, -2, -1)


**Задача 2** (25 баллов)

Найдите условные экстремумы функции:

$$
f(x, y)=4 x+8 y, y^2-2 x y+5=0
$$

Вычислите результат самостоятельно (вручную) и с помощью Python. Сравните результаты.

Функция Лагранжа:
$\mathcal{L}(x, y, \lambda) = 4x + 8y + \lambda (y^2 - 2xy + 5)$

Уравнения Лагранжа:

$\frac{\partial \mathcal{L}}{\partial x} = 4 - 2\lambda y = 0$

$\frac{\partial \mathcal{L}}{\partial y} = 8 + 2\lambda y - 2\lambda x = 0$

$\frac{\partial \mathcal{L}}{\partial \lambda} = y^2 - 2xy + 5 = 0$

Решение системы:

1. Из первого уравнения:
   $\lambda = \frac{2}{y}$

2. Подставим $\lambda$ во второе уравнение:
   $8 + 2\left(\frac{2}{y}\right)(y - x) = 0$

   Упростим:
   $8 + \frac{4(y - x)}{y} = 0$

   
   Умножим на $y$ и раскроем скобки:
   $8y + 4y - 4x = 0 \implies x = 3y$

4. Подставим $x = 3y$ в ограничение:
 
   $y^2 - 2(3y)y + 5 = 0$

   $y^2 - 6y^2 + 5 = 0$

   
   $-5y^2 + 5 = 0 \implies y^2 = 1 \implies y = \pm 1$

6. Найдём $x$:
   - $y = 1 \implies x = 3 \cdot 1 = 3$
   - $y = -1\implies x = 3 \cdot (-1) = -3$

Ответ: точки условного экстремума: (3, 1), (-3, -1)


In [13]:
# переменные
x, y, l = sp.symbols('x, y, l')

# функция
f = 4 * x + 8 * y

# функция ограничений
f_c = y**2 - 2 * x * y + 5

# функция Лагранжа
L = f + l*f_c

# частные производные по трем переменным
L_dx = L.diff(x)
L_dy = L.diff(y)
L_dl = L.diff(l)

# приравниваем функции частных производных к 0 и решаем систему уравнений 
points = sp.solve([L_dx, L_dy, L_dl], [x, y, l])
print(f'Точки условного экстремума: {points[0][:-1]} и {points[1][:-1]}')

Точки условного экстремума: (-3, -1) и (3, 1)


Решение вручную и прогаммно дало одинаковый результат: точками условного экстремума являются (3, 1) и (-3, -1)

**Задача 3** (30 баллов)

Вам предложены данные с информацией о успеваемости студентов:

Независимые переменные:
* Hours Studied: Общее количество часов, потраченных на учебу каждым студентом.
* Previous Scores: Баллы, полученные студентами на предыдущих экзаменах.
* Sleep Hours: Среднее количество часов сна студента в сутки.
* Sample Question Papers Practiced: Количество пробных экзаменационных работ, с которыми студент занимался.
  
Целевая переменная:
* Performance Index: Показатель общей успеваемости каждого студента. Индекс успеваемости отражает академическую успеваемость студента и округляется до ближайшего целого числа. Индекс варьируется от 10 до 100, при этом более высокие значения свидетельствуют о более высокой успеваемости.

**Решите задачу линейной регрессии, реализовав градиентный спуск самостоятельно, не используя готовое решение из какой-либо библиотеки.**

In [14]:
data = pd.read_csv('Student_Performance.csv')
# удаление не указанного в задании столбца
data.drop('Extracurricular Activities', axis=1, inplace=True)

# количество наблюдений
n = data.shape[0]

display(data.describe())
display(data.info())

Hours Studied  Previous Scores   Sleep Hours  \
count   10000.000000     10000.000000  10000.000000   
mean        4.992900        69.445700      6.530600   
std         2.589309        17.343152      1.695863   
min         1.000000        40.000000      4.000000   
25%         3.000000        54.000000      5.000000   
50%         5.000000        69.000000      7.000000   
75%         7.000000        85.000000      8.000000   
max         9.000000        99.000000      9.000000   

       Sample Question Papers Practiced  Performance Index  
count                      10000.000000       10000.000000  
mean                           4.583300          55.224800  
std                            2.867348          19.212558  
min                            0.000000          10.000000  
25%                            2.000000          40.000000  
50%                            5.000000          55.000000  
75%                            7.000000          71.000000  
max                            9.000000         100.000000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Hours Studied                     10000 non-null  int64  
 1   Previous Scores                   10000 non-null  int64  
 2   Sleep Hours                       10000 non-null  int64  
 3   Sample Question Papers Practiced  10000 non-null  int64  
 4   Performance Index                 10000 non-null  float64
dtypes: float64(1), int64(4)
memory usage: 390.8 KB


None

1. В данных отсутствуют пропуски
2. Один из признаков имеет тип с плавающей точкой, тогда как остальные - целые
3. Все признаки числовые

In [15]:
# признаки
X = data[['Hours Studied', 'Previous Scores', 'Sleep Hours', 'Sample Question Papers Practiced']].values.reshape(n, 4)
# целевой вектор
y = data['Performance Index'].values.reshape(n, 1)

# стандартизация признаков
scaler = StandardScaler()
X = scaler.fit_transform(X)

# добавление интерцепта
X = np.append(np.ones((n, 1)), X, axis = 1)

# начальный вектор коэффициентов
w = np.zeros((5,1))

# функция потерь
def cost_function(X, y , par):
    y_pred = X @ par
    error = (y_pred - y)**2
    cost = 1 / n * np.sum(error)
    return cost
    
# градиентный спуск
def grad_d(X, y, par, alpha, iterations, eps=0.001):
    costs = []
    for i in range(iterations):
        y_pred = X @ par
        der = np.dot(X.transpose(), (y_pred - y)) / n
        par -= alpha * der
        costs.append(cost_function(X, y, par))
        if np.linalg.norm(der) < eps:
            break
    return par, costs
    
# вычисление коэффициентов регрессии
w, costs = grad_d(X,y, w, 0.00005, 500000)
print(w.round(3))

[[55.224]
 [ 7.388]
 [17.664]
 [ 0.808]
 [ 0.56 ]]


In [16]:
# предсказание с коэффициентами регрессии
y_pred = X @ w
# ошибка предсказания
mape = np.mean(abs(y-y_pred) / y)

In [17]:
# использование sklearn
# создание модели
model = LinearRegression()
# вычисление коэффициентов sklearn
model.fit(X,y)
# предсказание sklearn
y_pred_skl = model.predict(X)
# ошибка предсказания sklearn
mape_skl = np.mean(abs(y-y_pred) / y)

In [18]:
# создадим справочник из данных полученных с помощью ручной реализации и с помощью библиотеки sklearn
dict = {
    'Параметр': ['Коэффициент 0', 'Коэффициент 1', 'Коэффициент 2', 'Коэффициент 3', 'Коэффициент 4', 'MAPE'],
    'Ручная реализация': [
        np.round(w[0], 3), 
        np.round(w[1], 3), 
        np.round(w[2], 3), 
        np.round(w[3], 3), 
        np.round(w[4], 3), 
        np.round(mape, 3)
    ],
    'sklearn': [
        np.round(model.intercept_, 3), 
        np.round(model.coef_[0,1], 3), 
        np.round(model.coef_[0,2], 3), 
        np.round(model.coef_[0,3], 3), 
        np.round(model.coef_[0,4], 3), 
        np.round(mape_skl, 3)
    ]
}
# на основе справочника создадим датафрейм
df = pd.DataFrame(dict).set_index('Параметр').T.reset_index()
print(df.to_string(index=False))

            index Коэффициент 0 Коэффициент 1 Коэффициент 2 Коэффициент 3 Коэффициент 4   MAPE
Ручная реализация      [55.224]       [7.388]      [17.664]       [0.808]        [0.56]  0.035
          sklearn      [55.225]         7.388        17.665         0.808          0.56  0.035


Коэффициенты идентичны (c точностью 1e-3). Ошибка MAPE = 3.5% довольно низкая, а значит модель довольно точная